### PR TITLE
Asynchronous writes

### PR DESCRIPTION
**Pull request by BK1603**
Branch for implementing asynchronous writes in neovim.

### CODE REVIEW

**Comment by BK1603** [comment]
@teto, all the commmits from my precious PR were also added, can this be avoided?

**Comment by teto** [comment]
prefix the pr with [WIP]. Yes you can create another branch and `git cherry-pick` a range of commits.

**Comment by BK1603** [comment]
> Yes you can create another branch and `git cherry-pick` a range of commits.

I will fix it and make another PR asap.

**Comment by BK1603** [comment]
The code is working, I have to add the error display parts, I can either add it to backup_utils or move those functions back to fileio.c

**Comment by BK1603** [comment]
There's another thing that needs to be done, since for restoring backups, we would need the backup name, I can make the backup global in the backup_utils.c file and use the restore function over there. That's what I have thought of for now. 
What are your opinions on this?

**Comment by BK1603** [comment]
@teto, Well, I will be working on the error display part and the restore function now. What would then remain would be the final write function and calling them all asynchronously.

File: src/nvim/fileio.c [hunk 1]
```diff
@@ -2736,156 +2736,172 @@ buf_write(
...
-              != 0) {
-            SET_ERRMSG(_("E506: Can't write to backup file "
-                         "(add ! to override)"));
-          }
-
-#ifdef UNIX
-          os_file_settime((char *)backup,
-                          file_info_old.stat.st_atim.tv_sec,
-                          file_info_old.stat.st_mtim.tv_sec);
-#endif
```
Comment (BK1603) [inline src/nvim/fileio.c:2866]: @teto, A question:

Over here, if we fail copying the original file, we still try to set atime and mtime on the backup file. Is that intentional? What is the point of this if there is no file yet? If this isn't the case then I must have misunderstood something.

**Comment by BK1603** [comment]
@teto, for the async write, should I use an async flag in the `buf_write` function and use it to call the functions asynchronously? That could be done either by making more functions for `buf_write` and queuing them (using `uv_work_queue` or making a wrapper function that simply calls `buf_write` asynchronously if async flag is set. What would you suggest?

IMO, just creating a wrapper around `buf_write` to call it asynchronously would be a bit 'hacky', but doing that is pretty easy and there are very little chances of different behavior. Creating more functions and breaking down `buf_write` further would help with refactoring, but it would take some time.

File: src/nvim/fileio.c (hunk 1 above)
Comment (teto) [reply src/nvim/fileio.c:2866]: It does seem weird. I tried a git blame on neovim source code but with little success. Checking against the vim source code (grep for buf_write), it seems to do the same (does not check against vim_rename success). This seems wrong.

**Comment by teto** [comment]
we can adjust the API later, make sure the behavior is correct first.

**Comment by BK1603** [comment]
@teto, a status update. I am making functions for asynchronous call right now. I will push as soon as a few of them are callable from a new thread. Also I am looking for better ways of implementing this if possible, since it still isn't too late for that.

Upon trying to implement my approach of libuv work queues I realized that a huge baton would have to be passed around between multiple threads and the entire write function would be executed partly via the functions on a separate thread and partly via the callbacks passed to the work queue.
I currently am working on the callbacks.

Also I think I should use a new loop for this?

**Comment by teto** [comment]
what do you mean a `huge baton` ?

> Also I think I should use a new loop for this?

Is the added complexity worth the gain ? we need to make sure what is saved is the content of the buffer when starting to save. Make sure to write some tests if needed and iterate while keeping the tests green.

**Comment by BK1603** [comment]
> what do you mean a `huge baton` ?
> 
@teto, by the baton I mean the struct that must be maintained for transferring necessary values in
between threads. I think it can be reduced by a bit of careful in place calculation but I am not yet too sure of how much can it be reduced.

> 
> Is the added complexity worth the gain ? we need to make sure what is saved is the content of the buffer when starting to save. Make sure to write some tests if needed and iterate while keeping the tests green.

I'll make sure to write a few tests, also for the loop. In which loop should I queue these functions then? I thought having a separate loop for this would work better. I don't think it should be too complex. Just a `write_loop` containing the work queue for the relevant functions? But I might be misunderstanding something regarding the complexity.

**Comment by BK1603** [comment]
@teto, here is a broad overview of the current implementation plan:
First, make a backup asynchronously using a new thread. If the backup is successful, initialize `write_info` and call write. If the backup is not successful, display the error messages and return.
If the backup was successful, write should be called in a new thread. If the write is successful, do whatever we do at the end of write and return. If write isn't successful, restore the backup, display the error messages and return.

This implementation plan requires a struct to pass things around in various threads. Which is what I referred to as the `baton`. There are some other complications arising in implementing this. The biggest being separating the code into several callable functions that can be called asynchronously.  And then there is the question for which loop to use for these functions? I guess we cannot block the main event loop during the I/O, and honestly I am not aware of all the loops being used in neovim. I do know about the existence of the `fs_loop`, but I think blocking that would mean that we cannot read a new file until the old file is written.

So I am open to any suggestions regarding the loop or even the implementation plan.

**Comment by justinmk** [comment]
Looks like this is very much WIP

**Pull request closed without merging**